tests/snapshot_test.cc

```diff
@@ -0,0 +1,191 @@
+
+    std::filesystem::path GetTestTempDir() { return record_test_temp_dir_; }
+
+    std::filesystem::path GetSnapshotTestTempDir() { return snapshot_test_temp_dir_; }
+
+    RecorderConfig GetTestConfig();
+
+    void TestMakeSnapshots();
+
+    void TestSnapshotNum();
+
+    void TestSnapshotContent();
+
+   private:
+    const std::string     SnapshotDirName    = std::string("Snapshot");
+    const std::string     SnapshotSubdirName = std::string("SECONDLY");
+    std::filesystem::path record_test_temp_dir_{
+        std::filesystem::temp_directory_path() / (std::string("CRSnapshotTestTmpDir.") + std::to_string(getpid()))};
+    std::filesystem::path snapshot_test_temp_dir_{record_test_temp_dir_ / SnapshotDirName / SnapshotSubdirName};
+    std::filesystem::path record_dir_;
+    std::map<std::string, std::size_t> snapshot_made_map_;
+
+    static constexpr std::size_t     kThreadNum            = 4;
+    static constexpr std::size_t     kMessageNum           = 4;
+    static constexpr auto            kSleepBetweenMessages = std::chrono::seconds(1);
+    static constexpr channel_subid_t kTestIntChannelSubId  = 11;
+};
+
+template<class T>
+    requires std::is_standard_layout_v<T>
+struct TestMessage : public CRMessage<TestMessage<T>> {
+    using data_type = T;
+
+    TestMessage() = default;
+
+    explicit TestMessage(T val) : value_(std::move(val)) {}
+
+    T value_{};
+};
+
+template<class T>
+void MessageFromStr(TestMessage<T>& msg, const std::string& serialized_msg) {
+    std::memcpy(&msg.value_, serialized_msg.c_str(), std::min(sizeof(T), serialized_msg.size()));
+}
+
+template<class T>
+std::string MessageToStr(const TestMessage<T>& msg) {
+    std::string serialized_msg(sizeof(T), 0);
+    std::memcpy(&serialized_msg[0], &msg.value_, serialized_msg.size());
+    return serialized_msg;
+}
+
+RecorderConfig SnapshotTest::GetTestConfig() {
+    RecorderConfig::IntervalConfig int_config{
+        .name_         = SnapshotSubdirName,
+        .interval_sec_ = kSleepBetweenMessages,
+    };
+
+    return RecorderConfig{
+        .snapshot_intervals_ = {int_config},
+        .record_dir_         = GetTestTempDir(),
+    };
+}
+
+void SnapshotTest::TestMakeSnapshots() {
+    JobRunner::Config config = {
+        .thread_num_ = kThreadNum,
+    };
+    auto            runner = JobRunner::MakeJobRunner(config);
+    MessageRecorder recorder(GetTestConfig(), runner);
+
+    recorder.RegisterChannel<TestMessage<int>>(kTestIntChannelSubId);
+    record_dir_ = recorder.GetRecordDir();
+
+    core::CRNode publisher;
+
+    for (std::size_t i = 0; i <= kMessageNum; ++i) {
+        auto test_message = std::make_shared<TestMessage<int>>(static_cast<int>(i));
+        publisher.Publish(kTestIntChannelSubId, std::move(test_message));
+        snapshot_made_map_[fmt::format("{:%Y%m%d-%H%M%S.%Z}", std::chrono::system_clock::now())] = i;
+        std::this_thread::sleep_for(kSleepBetweenMessages);
```

Comment:
Looks like something's wrong here with the sleep method causing the `tsan` failure, the `recorder` seems to be fine